BUG: safer bincount casting

tylerjereddy opened the pull request:
* Fixes #28354

* Allow some uint types to be cast a bit more gracefully for `np.bincount()`.

tylerjereddy committed:
BUG: safer bincount casting

* Fixes #28354

* Force usage of `npy_intp` type in `np.bincount()` and avoid
unsafe casting errors with i.e., `npy_uint64`. This is similar
to our behavior with indexing.

tylerjereddy commented:
The CI failures for the new test case I added are real; `npy_intp` gets used in other places in the control flow as well, so patching locally to account for more explicit type casts alone isn't sufficient.

tylerjereddy commented:
Briefly discussed at triage meeting this morning--Sebastian suggested forcing a cast to `npy_intp`, even for `uint64` input, since indexing does that as well and it isn't really practical to guard against the unreasonably large ints that would wrap to negative values.

tylerjereddy commented on numpy/_core/src/multiarray/compiled_base.c:
```diff
@@ -177,6 +177,11 @@ arr_bincount(PyObject *NPY_UNUSED(self), PyObject *const *args,
+        list = PyArray_FromAny(list, local_dtype, 0, 0, NPY_ARRAY_FORCECAST, NULL);
+        if (list == NULL) {
+            goto fail;
+        }
```
Neither of the two CI failures look related and I checked that the new regression test indeed fails before/passes after the patch.

I suspect Nathan or Sebastian will have to babysit me on the reference counting/setting for `local_dtype` and/or `list` above perhaps. Anyway, happy to make whatever revisions are requested.

For coercing to `NPY_INTP` we also have `coerce_index` in `iterators.c`, but dropping that in seemed to potentially require a bit more work.

seberg replied:
Thanks, there are problems here, but the refcounting is fine :).

1. We shouldn't force-cast everything.  We can get away with force-casting (with some pragmatism) only for integer arrays.  So basically, the logic should be: `PyArray_ISINTEGER(list) ? unsafe-cast : safe-cast` (or similar)
2. `PyArray_FromAny` needs the flags to be passed because the later code relies on them.  I.e. we do need the "Coniguous" part here.  That is, it uses `NPY_ARRAY_DEFAULT` (it is just a one line macro for calling `PyArray_FromAny`).
   (Which after a few unpacking, translates to: contiguous, aligned and writeable.)

The current version should crash hard if you do something like `np.bincount(np.arange(10000)[::2])` and it is a bit scary that we have no test for that!

seberg replied:
(Ah, I spoke too fast, the refcounting is indeed not fine, as user input `list` is a borrowed reference normally and you are assigning a strong one and you need to `DECREF` it.  But once you refactor things, that problem probably goes away by itself.)

tylerjereddy committed:
MAINT, BUG: PR 28355 revisions

* `arr_bincount()` now only uses unsafe casting for integer
input types, and the number of casting operations has
been reduced for the code path used in above PR.

* a test for non-crashing behavior with non-contiguous
`bincount()` input has been added.

tylerjereddy commented:
> We shouldn't force-cast everything. 

Ok, I added a guard for that.

> PyArray_FromAny needs the flags to be passed because the later code relies on them..

I couldn't replicate the need for this because what I did originally was two casting operations instead of one--the `Any` cast was added in the control flow to precede the `PyArray_ContiguousFromAny` after it, mostly because I was being lazy about constructing the appropriate requirements/flags. In any case, I've simplified things to restore to a single casting operation now, with custom `flags` depending on input type, which I suspect is closer to what you'd want?

> The current version should crash hard if you do something like np.bincount(np.arange(10000)[::2])

I couldn't find any evidence of a hard crash with that even when I intentionally made things worse with regards to C contiguity, though it may require i.e., `pytest-repeat` and some patience. In any case, I've added that test case as a "doesn't segfault" test just in case.

tylerjereddy committed:
MAINT, BUG: PR 28355 revisions

* Based on reviewer feedback, narrow the scope of the `flags`
variable in `arr_bincount()`.

* Based on reviewer feedback, add an array-like test for the `uint64`
casting issue, which indeed fails before and passes after adding
a similar shim to the array-like code path in `arr_bincount()`.

* Based on reviewer feedback, switch the patching from `PyArray_Size`
to `PyArray_Check` in a few places.

tylerjereddy commented:
@seberg I think I've addressed most of the commits in the latest commit (see commit message).

Covering the array-like case does add a bit of code duplication now I suppose. I didn't follow the empty `uint8` worry, but switched to `PyArray_Check` in any case since I believe I was indeed guessing a bit with `Size` (now in two places, though didn't check we really need it in both).

Note that the test for array-like did fail before/pass after the additional coercion shim just added (I also noticed that our error message for passing in a raw array-like class, rather than instance, to `bincount` seems terrible, but that's probably another matter).

I don't think I followed the deprecation-related concern--I thought I was just coercing for integer type inputs and it was the non-integer cases that were deprecated. It is a bit confusing though, because the deprecation message itself seems simple while the commented discussion a bit higher up seems more confusing...

seberg reviewed (approved):
Thanks, duplicating this seems indeed like the easiest path.

seberg committed:
Update numpy/_core/src/multiarray/compiled_base.c

seberg commented:
FWIW, I think this has been around forever and the issue was just a cupy test, I think.  So I don't care much for backporting, but I also don't think there is a problem with that..

jakirkham commented:
As `2.2.x` looks pretty active, went ahead and created a backport to `2.2.x` with PR: https://github.com/numpy/numpy/pull/28420

It looks like versions older than that are not seeing much activity. So didn't go further. Please let me know whether that would be worthwhile
